finat/tensorfiniteelement.py

```diff
@@ -120,6 +120,27 @@ def _tensorise(self, scalar_evaluation):
             )
         return result
 
+    @property
+    def dual_basis(self):
+
+        scalar = self._base_element
+        Q, points = scalar.dual_basis
+
+        if len(self._shape) == 2:
+            # Suppose the tensor element has shape (2, 4)
+            # These identity matrices may have difference sizes depending the shapes
+            # tQ = Q ⊗ 𝟙₂ ⊗ 𝟙₄
+            # This can be concretely realised with
+            indices = gem.indices(len(Q.shape))
+            i, j = (gem.Index(extent=self._shape[0]) for _ in range(self._shape[0]))
+            k, l = (gem.Index(extent=self._shape[1]) for _ in range(self._shape[1])) # noqa E741
+            # TODO Need to check how this plays with the transpose argument to TensorFiniteElement.
+            tQ = gem.ComponentTensor(Q[indices]*gem.Delta(i, j)*gem.Delta(k, l), indices + (i, j, k, l))
```

Comment:
There's something wrong here. This gives `tQ` with shape `[Q.shape[0], Q.shape[1], 2, 2, 4, 4]` where I would expect to get `tQ` with shape `[Q.shape[0], Q.shape[1], 2, 4]`

Reply:
Why? `CT(Delta(i, j))` has shape `(i.extent, j.extent)`

Reply:
I mean that the maths doesn't seem to add up. Ultimately I need to tensor contract this `tQ` with `expr = fn(points)` but it seems that my `tQ` has too much shape.

Reply:
I thought that the answer might be
```python
shape_indices = tuple(gem.Index(extent=ex) for ex in Q.shape)
            expansion_indices = tuple(gem.Index(extent=ex) for ex in self._shape)
            tQ = gem.ComponentTensor(Q[shape_indices]*gem.Delta(expansion_indices[0], expansion_indices[1]), shape_indices + expansion_indices )
```
but this seems to cause interpolation to give the wrong answers when compared to projection (in tests that ought to pass)

Reply:
I think you want one delta for each shape axis. So something like:
```
is = tuple(gem.Index(extent=d) for d in self._shape)
js = tuple(gem.Index(extent=d) for d in self._shape)

deltas = reduce(gem.Product, (gem.Delta(i, j) for i, j in zip(is, js)))
```

Reply:
> I mean that the maths doesn't seem to add up. Ultimately I need to tensor contract this `tQ` with `expr = fn(points)` but it seems that my `tQ` has too much shape.

Can you write it out? Section 2.3 of https://arxiv.org/pdf/1711.02473.pdf may be helpful

Reply:
Your `expr` should, I think, have shape `tensor_shape` at each point, is that right? Suppose it is `(2, 3)`. And there are `Q` points, so the total shape is `(Q, 2, 3)`. We index with: `q, i, j` to produce

```
exprqij = expr[q, i, j]
```
Now we contract over `i` and `j` against `tQ` which has indices `q, i, j, k, l` because it looks like
```
tQ = Q[q]*Delta(i, k)*Delta(j, l)
```

So we have:
```
tQexprqkl = IndexSum(tQ*exprqij, (i, j))
```
and if you wrap this up in a CT you get

```
evaluation = CT(tQexprqkl, (q, k, l))
```
which has shape `(Q, 2, 3)` as required?

Reply:
Q for a general FIAT element has shape `(num_nodes, num_points, dual_weight_shape[0], ..., weight_shape[n]) ` assuming the `dual_weight` (the `q`s in the relevant code) is not scalar so I think this may be a little more complicated than this

Reply:
You also have to deal with `compile_gem` sometimes dropping indices when `expr` is generated, e.g. when your expression is constant

Reply:
It's hard this!

Reply:
> Q for a general FIAT element has shape `(num_nodes, num_points, dual_weight_shape[0], ..., weight_shape[n]) ` assuming the `dual_weight` (the `q`s in the relevant code) is not scalar so I think this may be a little more complicated than this

I think this doesn't go against what I said. Suppose we make a `TensorFiniteElement(base, shape=(2, 3))`. Where `base` is intrinsically vector-valued (so has shape `(4,)` say). That means that any expression that we interpolate has shape `(4, 2, 3)` and, because the basis function is vector-valued, we want to produce a dof-value that has shape `(2, 3)`.

The `num_nodes` part is "easy" because that's an output axis. So for each node, you evaluate `expr` at some `nQ` points and get back an expression that has shape `(nQ, 4, 2, 3)`. The base element will advertise a dual that has shape (for each node) `(nQ, 4)`, I think. So we have:
```python
expr = ... # shape (nQ, 4, 2, 3)
base_dual = ... # shape `(nQ, 4)`
q = gem.Index()
b, = base.get_value_indices()
i, j = (gem.Index(extent=d) for d in (2, 3)) # generally self._shape
k, l = (gem.Index(extent=d) for d in (2, 3))
deltas = gem.Delta(i, k)*gem.Delta(j, l)
dual = base_dual[q, b]*deltas # no shape, free indices q, b, i, j, k, l
expr = expr[q, b, i, j]
to_contract = dual*expr # no shape, free indices q, b, i, j, k, l
contracted = gem.IndexSum(to_contract, (q, b, i, j)) # no shape, free indices k, l
result = gem.ComponentTensor(contracted, (k, l)) # shape (2, 3), no free indices
```

Is that right?

Reply:
> I think you want one delta for each shape axis. So something like:
> 
> ```
> is = tuple(gem.Index(extent=d) for d in self._shape)
> js = tuple(gem.Index(extent=d) for d in self._shape)
> 
> deltas = reduce(gem.Product, (gem.Delta(i, j) for i, j in zip(is, js)))
> ```



> Your `expr` should, I think, have shape `tensor_shape` at each point, is that right? Suppose it is `(2, 3)`. And there are `Q` points, so the total shape is `(Q, 2, 3)`. We index with: `q, i, j` to produce
> 
> ```
> exprqij = expr[q, i, j]
> ```
> 
> Now we contract over `i` and `j` against `tQ` which has indices `q, i, j, k, l` because it looks like
> 
> ```
> tQ = Q[q]*Delta(i, k)*Delta(j, l)
> ```
> 
> So we have:
> 
> ```
> tQexprqkl = IndexSum(tQ*exprqij, (i, j))
> ```
> 
> and if you wrap this up in a CT you get
> 
> ```
> evaluation = CT(tQexprqkl, (q, k, l))
> ```
> 
> which has shape `(Q, 2, 3)` as required?

Okay I think combining these two suggestions is basically right, although you have to be careful to replace the relevant of the `q` indices (which you might have gotten from the shape of `Q`) with with the indices from the point set you contract over (typically the second index though it depends if it's the expression which adds shape or the shape of the points themselves).